// @delon/theme/__ivy_ngcc__/fesm2015/layout-default.js
import { DOCUMENT, CommonModule } from '@angular/common';
import { Component, ViewChild, Input, ElementRef, Renderer2, Inject, ContentChildren, Directive, ChangeDetectionStrategy, ChangeDetectorRef, EventEmitter, ViewEncapsulation, NgZone, Optional, Output, NgModule } from '@angular/core';
import { RouteConfigLoadStart, NavigationError, NavigationCancel, NavigationEnd, RouteConfigLoadEnd, Router, RouterModule } from '@angular/router';
import { Subject } from 'rxjs';
import { takeUntil, filter } from 'rxjs/operators';
import { SettingsService, MenuService } from '@delon/theme';
import { updateHostClass } from '@delon/util/browser';
import { NzMessageService, NzMessageModule } from 'ng-zorro-antd/message';
import { NzAvatarModule } from 'ng-zorro-antd/avatar';
import { NzDropDownModule } from 'ng-zorro-antd/dropdown';
import { NzIconModule } from 'ng-zorro-antd/icon';
import { NzToolTipModule } from 'ng-zorro-antd/tooltip';
import { __decorate } from 'tslib';
import { Directionality } from '@angular/cdk/bidi';
import { DomSanitizer } from '@angular/platform-browser';
import { InputBoolean, InputNumber, ZoneOutside } from '@delon/util/decorator';
import { WINDOW } from '@delon/util/token';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from 'ng-zorro-antd/message';
import * as ɵngcc3 from '@delon/theme';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from 'ng-zorro-antd/icon';
import * as ɵngcc6 from '@angular/platform-browser';
import * as ɵngcc7 from '@angular/cdk/bidi';
import * as ɵngcc8 from 'ng-zorro-antd/tooltip';

const _c0 = ["host"];
function LayoutDefaultHeaderItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c1 = ["*"];
function LayoutDefaultComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 4);
} }
function LayoutDefaultComponent_div_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function LayoutDefaultComponent_div_2_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function LayoutDefaultComponent_div_2_layout_default_nav_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "layout-default-nav", 8);
} }
function LayoutDefaultComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵelementStart(1, "div", 6);
    ɵngcc0.ɵɵtemplate(2, LayoutDefaultComponent_div_2_ng_container_2_Template, 1, 0, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, LayoutDefaultComponent_div_2_ng_container_3_Template, 1, 0, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(4, LayoutDefaultComponent_div_2_layout_default_nav_4_Template, 1, 0, "layout-default-nav", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.asideUser);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.nav);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.nav);
} }
function LayoutDefaultComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function LayoutDefaultHeaderComponent_ng_template_0_li_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function LayoutDefaultHeaderComponent_ng_template_0_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultHeaderComponent_ng_template_0_li_0_ng_container_1_Template, 1, 0, "ng-container", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r8 = ctx.$implicit;
    ɵngcc0.ɵɵclassProp("hidden-mobile", i_r8.hidden === "mobile")("hidden-pc", i_r8.hidden === "pc");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", i_r8.host);
} }
function LayoutDefaultHeaderComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, LayoutDefaultHeaderComponent_ng_template_0_li_0_Template, 2, 5, "li", 10);
} if (rf & 2) {
    const ls_r6 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngForOf", ls_r6);
} }
function LayoutDefaultHeaderComponent_li_8_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵelementStart(1, "div", 12);
    ɵngcc0.ɵɵlistener("click", function LayoutDefaultHeaderComponent_li_8_Template_div_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.toggleCollapsed(); });
    ɵngcc0.ɵɵelement(2, "i", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("nzType", ctx_r2.collapsedIcon);
} }
function LayoutDefaultHeaderComponent_ng_template_9_Template(rf, ctx) { }
function LayoutDefaultHeaderComponent_div_10_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function LayoutDefaultHeaderComponent_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultHeaderComponent_div_10_ng_container_1_Template, 1, 0, "ng-container", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r4.middle[0].host);
} }
function LayoutDefaultHeaderComponent_ng_template_12_Template(rf, ctx) { }
const _c2 = function (a0) { return { $implicit: a0 }; };
function LayoutDefaultNavComponent_ng_template_0_ng_container_0_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 10);
} if (rf & 2) {
    const i_r5 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("nzType", i_r5.value)("nzTheme", i_r5.theme)("nzSpin", i_r5.spin)("nzTwotoneColor", i_r5.twoToneColor)("nzIconfont", i_r5.iconfont)("nzRotate", i_r5.rotate);
} }
function LayoutDefaultNavComponent_ng_template_0_ng_container_0_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 11);
} if (rf & 2) {
    const i_r5 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("nzIconfont", i_r5.iconfont);
} }
function LayoutDefaultNavComponent_ng_template_0_ng_container_0_img_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 12);
} if (rf & 2) {
    const i_r5 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("src", i_r5.value, ɵngcc0.ɵɵsanitizeUrl);
} }
function LayoutDefaultNavComponent_ng_template_0_ng_container_0_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i");
} if (rf & 2) {
    const i_r5 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵclassMapInterpolate1("sidebar-nav__item-icon ", i_r5.value, "");
} }
function LayoutDefaultNavComponent_ng_template_0_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0, 5);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_0_ng_container_0_i_1_Template, 1, 6, "i", 6);
    ɵngcc0.ɵɵtemplate(2, LayoutDefaultNavComponent_ng_template_0_ng_container_0_i_2_Template, 1, 1, "i", 7);
    ɵngcc0.ɵɵtemplate(3, LayoutDefaultNavComponent_ng_template_0_ng_container_0_img_3_Template, 1, 1, "img", 8);
    ɵngcc0.ɵɵtemplate(4, LayoutDefaultNavComponent_ng_template_0_ng_container_0_i_4_Template, 1, 3, "i", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngSwitch", i_r5.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "icon");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "iconfont");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "img");
} }
function LayoutDefaultNavComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, LayoutDefaultNavComponent_ng_template_0_ng_container_0_Template, 5, 4, "ng-container", 4);
} if (rf & 2) {
    const i_r5 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", i_r5);
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_ng_container_1_ng_template_1_Template(rf, ctx) { }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 23);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r18 = ɵngcc0.ɵɵnextContext(4).$implicit;
    ɵngcc0.ɵɵnextContext(2);
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c2, i_r18.icon));
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_span_2_ng_template_1_Template(rf, ctx) { }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 24);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_span_2_ng_template_1_Template, 0, 0, "ng-template", 23);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r18 = ɵngcc0.ɵɵnextContext(4).$implicit;
    ɵngcc0.ɵɵnextContext(2);
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵproperty("nzTooltipTitle", i_r18.text);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c2, i_r18.icon));
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_ng_container_1_Template, 2, 4, "ng-container", 20);
    ɵngcc0.ɵɵtemplate(2, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_span_2_Template, 2, 5, "span", 22);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r24 = ɵngcc0.ɵɵnextContext(5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r24.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r24.collapsed);
} }
const _c3 = function (a0) { return { "sidebar-nav__item-disabled": a0 }; };
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r33 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 19);
    ɵngcc0.ɵɵlistener("click", function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r33); const i_r18 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r31 = ɵngcc0.ɵɵnextContext(2); return ctx_r31.to(i_r18); });
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_ng_container_1_Template, 3, 2, "ng-container", 20);
    ɵngcc0.ɵɵelement(2, "span", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r18 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c3, i_r18.disabled));
    ɵngcc0.ɵɵattribute("data-id", i_r18._id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r18._needIcon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", i_r18._text, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵattribute("title", i_r18.text);
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_2_ng_template_1_Template(rf, ctx) { }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r38 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 25);
    ɵngcc0.ɵɵlistener("click", function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r38); const i_r18 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r36 = ɵngcc0.ɵɵnextContext(2); return ctx_r36.toggleOpen(i_r18); })("mouseenter", function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_2_Template_a_mouseenter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const i_r18 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r39 = ɵngcc0.ɵɵnextContext(2); return ctx_r39.showSubMenu($event, i_r18); });
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_2_ng_template_1_Template, 0, 0, "ng-template", 23);
    ɵngcc0.ɵɵelement(2, "span", 21);
    ɵngcc0.ɵɵelement(3, "i", 26);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r18 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵnextContext(2);
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(4, _c2, i_r18.icon));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", i_r18._text, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵattribute("title", i_r18.text);
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "em");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r18 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵclassMapInterpolate1("badge badge-", i_r18.badgeStatus, "");
    ɵngcc0.ɵɵclassProp("badge-dot", i_r18.badgeDot);
    ɵngcc0.ɵɵattribute("title", i_r18.badge);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(i_r18.badge);
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_ul_4_ng_template_1_Template(rf, ctx) { }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_ul_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul");
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_ul_4_ng_template_1_Template, 0, 0, "ng-template", 23);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r18 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵnextContext(2);
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵclassMapInterpolate1("sidebar-nav sidebar-nav__sub sidebar-nav__depth", i_r18._depth, "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(5, _c2, i_r18.children));
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 14);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_1_Template, 3, 7, "a", 15);
    ɵngcc0.ɵɵtemplate(2, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_a_2_Template, 4, 6, "a", 16);
    ɵngcc0.ɵɵtemplate(3, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_div_3_Template, 3, 7, "div", 17);
    ɵngcc0.ɵɵtemplate(4, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_ul_4_Template, 2, 7, "ul", 18);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r18 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵclassProp("sidebar-nav__selected", i_r18._selected)("sidebar-nav__open", i_r18._open);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r18.children.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r18.children.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r18.badge);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r18.children.length > 0);
} }
function LayoutDefaultNavComponent_ng_template_2_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_template_2_ng_container_0_li_1_Template, 5, 8, "li", 13);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r18 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", i_r18._hidden !== true);
} }
function LayoutDefaultNavComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, LayoutDefaultNavComponent_ng_template_2_ng_container_0_Template, 2, 1, "ng-container", 3);
} if (rf & 2) {
    const ls_r16 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngForOf", ls_r16);
} }
function LayoutDefaultNavComponent_ng_container_5_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 28);
    ɵngcc0.ɵɵelement(1, "span", 29);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const group_r46 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", group_r46._text, ɵngcc0.ɵɵsanitizeHtml);
} }
function LayoutDefaultNavComponent_ng_container_5_ng_template_2_Template(rf, ctx) { }
function LayoutDefaultNavComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LayoutDefaultNavComponent_ng_container_5_li_1_Template, 2, 1, "li", 27);
    ɵngcc0.ɵɵtemplate(2, LayoutDefaultNavComponent_ng_container_5_ng_template_2_Template, 0, 0, "ng-template", 23);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const group_r46 = ctx.$implicit;
    ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", group_r46.group);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c2, group_r46.children));
} }
class LayoutDefaultHeaderItemComponent {
    constructor() {
        this.hidden = 'none';
        this.direction = 'right';
    }
}
LayoutDefaultHeaderItemComponent.ɵfac = function LayoutDefaultHeaderItemComponent_Factory(t) { return new (t || LayoutDefaultHeaderItemComponent)(); };
LayoutDefaultHeaderItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LayoutDefaultHeaderItemComponent, selectors: [["layout-default-header-item"]], viewQuery: function LayoutDefaultHeaderItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.host = _t.first);
    } }, inputs: { hidden: "hidden", direction: "direction" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["host", ""]], template: function LayoutDefaultHeaderItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, LayoutDefaultHeaderItemComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
LayoutDefaultHeaderItemComponent.propDecorators = {
    host: [{ type: ViewChild, args: ['host', { static: true },] }],
    hidden: [{ type: Input }],
    direction: [{ type: Input }]
};

class LayoutDefaultComponent {
    constructor(router, msgSrv, settings, el, renderer, doc) {
        this.settings = settings;
        this.el = el;
        this.renderer = renderer;
        this.doc = doc;
        this.destroy$ = new Subject();
        this.isFetching = false;
        router.events.pipe(takeUntil(this.destroy$)).subscribe(evt => {
            if (!this.isFetching && evt instanceof RouteConfigLoadStart) {
                this.isFetching = true;
            }
            if (evt instanceof NavigationError || evt instanceof NavigationCancel) {
                this.isFetching = false;
                if (evt instanceof NavigationError) {
                    msgSrv.error(`Could not load ${evt.url} route`, { nzDuration: 1000 * 3 });
                }
                return;
            }
            if (!(evt instanceof NavigationEnd || evt instanceof RouteConfigLoadEnd)) {
                return;
            }
            if (this.isFetching) {
                setTimeout(() => {
                    this.isFetching = false;
                }, 100);
            }
        });
    }
    setClass() {
        const { el, doc, renderer, settings } = this;
        const layout = settings.layout;
        updateHostClass(el.nativeElement, renderer, {
            ['alain-default']: true,
            [`alain-default__fixed`]: layout.fixed,
            [`alain-default__collapsed`]: layout.collapsed,
            [`alain-default__hide-aside`]: this.options.hideAside
        });
        doc.body.classList[layout.colorWeak ? 'add' : 'remove']('color-weak');
    }
    ngOnInit() {
        this.options = Object.assign({ logoExpanded: `./assets/logo-full.svg`, logoCollapsed: `./assets/logo.svg`, logoLink: `/`, hideAside: false }, this.options);
        const { settings, destroy$ } = this;
        settings.notify.pipe(takeUntil(destroy$)).subscribe(() => this.setClass());
        this.setClass();
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
LayoutDefaultComponent.ɵfac = function LayoutDefaultComponent_Factory(t) { return new (t || LayoutDefaultComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.NzMessageService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.SettingsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(DOCUMENT)); };
LayoutDefaultComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LayoutDefaultComponent, selectors: [["layout-default"]], contentQueries: function LayoutDefaultComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, LayoutDefaultHeaderItemComponent, 4);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerItems = _t);
    } }, inputs: { options: "options", asideUser: "asideUser", nav: "nav", content: "content" }, exportAs: ["layoutDefault"], ngContentSelectors: _c1, decls: 6, vars: 3, consts: [["class", "alain-default__progress-bar", 4, "ngIf"], ["class", "alain-default__aside", 4, "ngIf"], [1, "alain-default__content"], [4, "ngTemplateOutlet"], [1, "alain-default__progress-bar"], [1, "alain-default__aside"], [1, "alain-default__aside-inner"], ["class", "d-block py-lg", 4, "ngIf"], [1, "d-block", "py-lg"]], template: function LayoutDefaultComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, LayoutDefaultComponent_div_0_Template, 1, 0, "div", 0);
        ɵngcc0.ɵɵelement(1, "layout-default-header");
        ɵngcc0.ɵɵtemplate(2, LayoutDefaultComponent_div_2_Template, 5, 3, "div", 1);
        ɵngcc0.ɵɵelementStart(3, "section", 2);
        ɵngcc0.ɵɵtemplate(4, LayoutDefaultComponent_ng_container_4_Template, 1, 0, "ng-container", 3);
        ɵngcc0.ɵɵprojection(5);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isFetching);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.options.hideAside);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.content);
    } }, directives: function () { return [ɵngcc4.NgIf, LayoutDefaultHeaderComponent, ɵngcc4.NgTemplateOutlet, LayoutDefaultNavComponent]; }, encapsulation: 2 });
LayoutDefaultComponent.ctorParameters = () => [
    { type: Router },
    { type: NzMessageService },
    { type: SettingsService },
    { type: ElementRef },
    { type: Renderer2 },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
LayoutDefaultComponent.propDecorators = {
    headerItems: [{ type: ContentChildren, args: [LayoutDefaultHeaderItemComponent, { descendants: false },] }],
    options: [{ type: Input }],
    asideUser: [{ type: Input }],
    nav: [{ type: Input }],
    content: [{ type: Input }]
};

class LayoutDefaultHeaderItemTriggerDirective {
}
LayoutDefaultHeaderItemTriggerDirective.ɵfac = function LayoutDefaultHeaderItemTriggerDirective_Factory(t) { return new (t || LayoutDefaultHeaderItemTriggerDirective)(); };
LayoutDefaultHeaderItemTriggerDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: LayoutDefaultHeaderItemTriggerDirective, selectors: [["", "layout-default-header-item-trigger", ""]], hostVars: 2, hostBindings: function LayoutDefaultHeaderItemTriggerDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("alain-default__nav-item", true);
    } } });

class LayoutDefaultHeaderComponent {
    constructor(settings, parent, cdr) {
        this.settings = settings;
        this.parent = parent;
        this.cdr = cdr;
        this.destroy$ = new Subject();
        this.left = [];
        this.middle = [];
        this.right = [];
    }
    get options() {
        return this.parent.options;
    }
    get app() {
        return this.settings.app;
    }
    get collapsed() {
        return this.settings.layout.collapsed;
    }
    get collapsedIcon() {
        let type = this.collapsed ? 'unfold' : 'fold';
        if (this.settings.layout.direction === 'rtl') {
            type = this.collapsed ? 'fold' : 'unfold';
        }
        return `menu-${type}`;
    }
    refresh() {
        const arr = this.parent.headerItems.toArray();
        this.left = arr.filter(i => i.direction === 'left');
        this.middle = arr.filter(i => i.direction === 'middle');
        this.right = arr.filter(i => i.direction === 'right');
        this.cdr.detectChanges();
    }
    ngAfterViewInit() {
        this.parent.headerItems.changes.pipe(takeUntil(this.destroy$)).subscribe(() => this.refresh());
        this.refresh();
    }
    toggleCollapsed() {
        this.settings.setLayout('collapsed', !this.settings.layout.collapsed);
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
LayoutDefaultHeaderComponent.ɵfac = function LayoutDefaultHeaderComponent_Factory(t) { return new (t || LayoutDefaultHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc3.SettingsService), ɵngcc0.ɵɵdirectiveInject(LayoutDefaultComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
LayoutDefaultHeaderComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LayoutDefaultHeaderComponent, selectors: [["layout-default-header"]], hostVars: 2, hostBindings: function LayoutDefaultHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("alain-default__header", true);
    } }, decls: 13, vars: 17, consts: [["render", ""], [1, "alain-default__header-logo"], [1, "alain-default__header-logo-link", 3, "routerLink"], [1, "alain-default__header-logo-expanded"], [1, "alain-default__header-logo-collapsed"], [1, "alain-default__nav-wrap"], [1, "alain-default__nav"], [4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["class", "alain-default__nav alain-default__nav-middle", 4, "ngIf"], [3, "hidden-mobile", "hidden-pc", 4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], [1, "alain-default__nav-item", "alain-default__nav-item--collapse", 3, "click"], ["nz-icon", "", 3, "nzType"], [1, "alain-default__nav", "alain-default__nav-middle"]], template: function LayoutDefaultHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, LayoutDefaultHeaderComponent_ng_template_0_Template, 1, 1, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵelementStart(3, "a", 2);
        ɵngcc0.ɵɵelement(4, "img", 3);
        ɵngcc0.ɵɵelement(5, "img", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵelementStart(7, "ul", 6);
        ɵngcc0.ɵɵtemplate(8, LayoutDefaultHeaderComponent_li_8_Template, 3, 1, "li", 7);
        ɵngcc0.ɵɵtemplate(9, LayoutDefaultHeaderComponent_ng_template_9_Template, 0, 0, "ng-template", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, LayoutDefaultHeaderComponent_div_10_Template, 2, 1, "div", 9);
        ɵngcc0.ɵɵelementStart(11, "ul", 6);
        ɵngcc0.ɵɵtemplate(12, LayoutDefaultHeaderComponent_ng_template_12_Template, 0, 0, "ng-template", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵstyleProp("width", ctx.options.logoFixWidth, "px");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("routerLink", ctx.options.logoLink);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("src", ctx.options.logoExpanded, ɵngcc0.ɵɵsanitizeUrl)("alt", ctx.app.name);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("src", ctx.options.logoCollapsed, ɵngcc0.ɵɵsanitizeUrl)("alt", ctx.app.name);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.options.hideAside);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(13, _c2, ctx.left));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.middle.length > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(15, _c2, ctx.right));
    } }, directives: [ɵngcc1.RouterLinkWithHref, ɵngcc4.NgIf, ɵngcc4.NgTemplateOutlet, ɵngcc4.NgForOf, ɵngcc5.NzIconDirective], encapsulation: 2, changeDetection: 0 });
LayoutDefaultHeaderComponent.ctorParameters = () => [
    { type: SettingsService },
    { type: LayoutDefaultComponent },
    { type: ChangeDetectorRef }
];

const SHOWCLS = 'sidebar-nav__floating-show';
const FLOATINGCLS = 'sidebar-nav__floating';
class LayoutDefaultNavComponent {
    constructor(menuSrv, settings, router, render, cdr, ngZone, sanitizer, doc, win, directionality) {
        this.menuSrv = menuSrv;
        this.settings = settings;
        this.router = router;
        this.render = render;
        this.cdr = cdr;
        this.ngZone = ngZone;
        this.sanitizer = sanitizer;
        this.doc = doc;
        this.win = win;
        this.directionality = directionality;
        this.destroy$ = new Subject();
        this.dir = 'ltr';
        this.list = [];
        this.disabledAcl = false;
        this.autoCloseUnderPad = true;
        this.recursivePath = true;
        this.openStrictly = false;
        this.maxLevelIcon = 3;
        this.select = new EventEmitter();
    }
    get collapsed() {
        return this.settings.layout.collapsed;
    }
    getLinkNode(node) {
        node = node.nodeName === 'A' ? node : node.parentNode;
        return node.nodeName !== 'A' ? null : node;
    }
    floatingClickHandle(e) {
        e.stopPropagation();
        const linkNode = this.getLinkNode(e.target);
        if (linkNode == null) {
            return false;
        }
        const id = +linkNode.dataset.id;
        // Should be ingore children title trigger event
        if (isNaN(id)) {
            return false;
        }
        let item;
        this.menuSrv.visit(this.list, (i) => {
            if (!item && i._id === id) {
                item = i;
            }
        });
        this.to(item);
        this.hideAll();
        e.preventDefault();
        return false;
    }
    clearFloating() {
        if (!this.floatingEl)
            return;
        this.floatingEl.removeEventListener('click', this.floatingClickHandle.bind(this));
        // fix ie: https://github.com/ng-alain/delon/issues/52
        if (this.floatingEl.hasOwnProperty('remove')) {
            this.floatingEl.remove();
        }
        else if (this.floatingEl.parentNode) {
            this.floatingEl.parentNode.removeChild(this.floatingEl);
        }
    }
    genFloating() {
        this.clearFloating();
        this.floatingEl = this.render.createElement('div');
        this.floatingEl.classList.add(`${FLOATINGCLS}-container`);
        this.floatingEl.addEventListener('click', this.floatingClickHandle.bind(this), false);
        this.bodyEl.appendChild(this.floatingEl);
    }
    genSubNode(linkNode, item) {
        const id = `_sidebar-nav-${item._id}`;
        const childNode = item.badge ? linkNode.nextElementSibling.nextElementSibling : linkNode.nextElementSibling;
        const node = childNode.cloneNode(true);
        node.id = id;
        node.classList.add(FLOATINGCLS);
        node.addEventListener('mouseleave', () => {
            node.classList.remove(SHOWCLS);
        }, false);
        this.floatingEl.appendChild(node);
        return node;
    }
    hideAll() {
        const allNode = this.floatingEl.querySelectorAll(`.${FLOATINGCLS}`);
        for (let i = 0; i < allNode.length; i++) {
            allNode[i].classList.remove(SHOWCLS);
        }
    }
    // calculate the node position values.
    calPos(linkNode, node) {
        const rect = linkNode.getBoundingClientRect();
        // bug: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/14721015/
        const scrollTop = Math.max(this.doc.documentElement.scrollTop, this.bodyEl.scrollTop);
        const docHeight = Math.max(this.doc.documentElement.clientHeight, this.bodyEl.clientHeight);
        const spacing = 5;
        let offsetHeight = -spacing;
        if (docHeight < rect.top + node.clientHeight) {
            offsetHeight = rect.top + node.clientHeight - docHeight + spacing;
        }
        node.style.top = `${rect.top + scrollTop - offsetHeight}px`;
        if (this.dir === 'rtl') {
            node.style.right = `${rect.width + spacing}px`;
        }
        else {
            node.style.left = `${rect.right + spacing}px`;
        }
    }
    showSubMenu(e, item) {
        if (this.collapsed !== true) {
            return;
        }
        e.preventDefault();
        const linkNode = e.target;
        this.genFloating();
        const subNode = this.genSubNode(linkNode, item);
        this.hideAll();
        subNode.classList.add(SHOWCLS);
        this.calPos(linkNode, subNode);
    }
    to(item) {
        this.select.emit(item);
        if (item.disabled)
            return;
        if (item.externalLink) {
            if (item.target === '_blank') {
                this.win.open(item.externalLink);
            }
            else {
                this.win.location.href = item.externalLink;
            }
            return;
        }
        this.ngZone.run(() => this.router.navigateByUrl(item.link));
    }
    toggleOpen(item) {
        if (!this.openStrictly) {
            this.menuSrv.visit(this.list, (i) => {
                if (i !== item)
                    i._open = false;
            });
            let pItem = item._parent;
            while (pItem) {
                pItem._open = true;
                pItem = pItem._parent;
            }
        }
        item._open = !item._open;
        this.cdr.markForCheck();
    }
    _click() {
        if (this.isPad && this.collapsed) {
            this.openAside(false);
            this.hideAll();
        }
    }
    _docClick() {
        if (this.collapsed) {
            this.hideAll();
        }
    }
    openedByUrl(url) {
        const { menuSrv, recursivePath, openStrictly } = this;
        let findItem = menuSrv.getHit(this.menuSrv.menus, url, recursivePath, (i) => {
            i._selected = false;
            if (!openStrictly) {
                i._open = false;
            }
        });
        if (findItem == null)
            return;
        do {
            findItem._selected = true;
            if (!openStrictly) {
                findItem._open = true;
            }
            findItem = findItem._parent;
        } while (findItem);
    }
    ngOnInit() {
        var _a;
        const { doc, router, destroy$, menuSrv, settings, cdr } = this;
        this.bodyEl = doc.querySelector('body');
        this.openedByUrl(router.url);
        this.ngZone.runOutsideAngular(() => this.genFloating());
        menuSrv.change.pipe(takeUntil(destroy$)).subscribe(data => {
            menuSrv.visit(data, (i, _p, depth) => {
                i._text = this.sanitizer.bypassSecurityTrustHtml(i.text);
                i._needIcon = depth <= this.maxLevelIcon && !!i.icon;
                if (!i._aclResult) {
                    if (this.disabledAcl) {
                        i.disabled = true;
                    }
                    else {
                        i._hidden = true;
                    }
                }
                if (this.openStrictly) {
                    i._open = i.open != null ? i.open : false;
                }
            });
            this.list = menuSrv.menus.filter((w) => w._hidden !== true);
            cdr.detectChanges();
        });
        router.events.pipe(takeUntil(destroy$)).subscribe(e => {
            if (e instanceof NavigationEnd) {
                this.openedByUrl(e.urlAfterRedirects);
                this.underPad();
                this.cdr.detectChanges();
            }
        });
        settings.notify
            .pipe(takeUntil(destroy$), filter(t => t.type === 'layout' && t.name === 'collapsed'))
            .subscribe(() => this.clearFloating());
        this.underPad();
        this.dir = this.directionality.value;
        (_a = this.directionality.change) === null || _a === void 0 ? void 0 : _a.pipe(takeUntil(destroy$)).subscribe((direction) => {
            this.dir = direction;
        });
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
        this.clearFloating();
    }
    // #region Under pad
    get isPad() {
        return this.doc.defaultView.innerWidth < 768;
    }
    underPad() {
        if (this.autoCloseUnderPad && this.isPad && !this.collapsed) {
            setTimeout(() => this.openAside(true));
        }
    }
    openAside(status) {
        this.settings.setLayout('collapsed', status);
    }
}
LayoutDefaultNavComponent.ɵfac = function LayoutDefaultNavComponent_Factory(t) { return new (t || LayoutDefaultNavComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc3.MenuService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.SettingsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.DomSanitizer), ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(WINDOW), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.Directionality, 8)); };
LayoutDefaultNavComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LayoutDefaultNavComponent, selectors: [["layout-default-nav"]], hostBindings: function LayoutDefaultNavComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function LayoutDefaultNavComponent_click_HostBindingHandler() { return ctx._click(); })("click", function LayoutDefaultNavComponent_click_HostBindingHandler() { return ctx._docClick(); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { disabledAcl: "disabledAcl", autoCloseUnderPad: "autoCloseUnderPad", recursivePath: "recursivePath", openStrictly: "openStrictly", maxLevelIcon: "maxLevelIcon" }, outputs: { select: "select" }, decls: 6, vars: 1, consts: [["icon", ""], ["tree", ""], [1, "sidebar-nav"], [4, "ngFor", "ngForOf"], [3, "ngSwitch", 4, "ngIf"], [3, "ngSwitch"], ["class", "sidebar-nav__item-icon", "nz-icon", "", 3, "nzType", "nzTheme", "nzSpin", "nzTwotoneColor", "nzIconfont", "nzRotate", 4, "ngSwitchCase"], ["class", "sidebar-nav__item-icon", "nz-icon", "", 3, "nzIconfont", 4, "ngSwitchCase"], ["class", "sidebar-nav__item-icon sidebar-nav__item-img", 3, "src", 4, "ngSwitchCase"], [3, "class", 4, "ngSwitchDefault"], ["nz-icon", "", 1, "sidebar-nav__item-icon", 3, "nzType", "nzTheme", "nzSpin", "nzTwotoneColor", "nzIconfont", "nzRotate"], ["nz-icon", "", 1, "sidebar-nav__item-icon", 3, "nzIconfont"], [1, "sidebar-nav__item-icon", "sidebar-nav__item-img", 3, "src"], ["class", "sidebar-nav__item", 3, "sidebar-nav__selected", "sidebar-nav__open", 4, "ngIf"], [1, "sidebar-nav__item"], ["class", "sidebar-nav__item-link", 3, "ngClass", "click", 4, "ngIf"], ["class", "sidebar-nav__item-link", 3, "click", "mouseenter", 4, "ngIf"], [3, "class", "badge-dot", 4, "ngIf"], [3, "class", 4, "ngIf"], [1, "sidebar-nav__item-link", 3, "ngClass", "click"], [4, "ngIf"], [1, "sidebar-nav__item-text", 3, "innerHTML"], ["nz-tooltip", "", "nzTooltipPlacement", "right", 3, "nzTooltipTitle", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["nz-tooltip", "", "nzTooltipPlacement", "right", 3, "nzTooltipTitle"], [1, "sidebar-nav__item-link", 3, "click", "mouseenter"], [1, "sidebar-nav__sub-arrow"], ["class", "sidebar-nav__item sidebar-nav__group-title", 4, "ngIf"], [1, "sidebar-nav__item", "sidebar-nav__group-title"], [3, "innerHTML"]], template: function LayoutDefaultNavComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, LayoutDefaultNavComponent_ng_template_0_Template, 1, 1, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(2, LayoutDefaultNavComponent_ng_template_2_Template, 1, 1, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementStart(4, "ul", 2);
        ɵngcc0.ɵɵtemplate(5, LayoutDefaultNavComponent_ng_container_5_Template, 3, 5, "ng-container", 3);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.list);
    } }, directives: [ɵngcc4.NgForOf, ɵngcc4.NgIf, ɵngcc4.NgSwitch, ɵngcc4.NgSwitchCase, ɵngcc4.NgSwitchDefault, ɵngcc5.NzIconDirective, ɵngcc4.NgClass, ɵngcc4.NgTemplateOutlet, ɵngcc8.NzTooltipDirective], encapsulation: 2, changeDetection: 0 });
LayoutDefaultNavComponent.ctorParameters = () => [
    { type: MenuService },
    { type: SettingsService },
    { type: Router },
    { type: Renderer2 },
    { type: ChangeDetectorRef },
    { type: NgZone },
    { type: DomSanitizer },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WINDOW,] }] },
    { type: Directionality, decorators: [{ type: Optional }] }
];
LayoutDefaultNavComponent.propDecorators = {
    disabledAcl: [{ type: Input }],
    autoCloseUnderPad: [{ type: Input }],
    recursivePath: [{ type: Input }],
    openStrictly: [{ type: Input }],
    maxLevelIcon: [{ type: Input }],
    select: [{ type: Output }]
};
__decorate([
    InputBoolean()
], LayoutDefaultNavComponent.prototype, "disabledAcl", void 0);
__decorate([
    InputBoolean()
], LayoutDefaultNavComponent.prototype, "autoCloseUnderPad", void 0);
__decorate([
    InputBoolean()
], LayoutDefaultNavComponent.prototype, "recursivePath", void 0);
__decorate([
    InputBoolean()
], LayoutDefaultNavComponent.prototype, "openStrictly", void 0);
__decorate([
    InputNumber()
], LayoutDefaultNavComponent.prototype, "maxLevelIcon", void 0);
__decorate([
    ZoneOutside()
], LayoutDefaultNavComponent.prototype, "showSubMenu", null);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LayoutDefaultHeaderItemComponent, [{
        type: Component,
        args: [{
                selector: 'layout-default-header-item',
                template: `
    <ng-template #host>
      <ng-content></ng-content>
    </ng-template>
  `
            }]
    }], function () { return []; }, { hidden: [{
            type: Input
        }], direction: [{
            type: Input
        }], host: [{
            type: ViewChild,
            args: ['host', { static: true }]
        }] }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LayoutDefaultComponent, [{
        type: Component,
        args: [{
                selector: 'layout-default',
                exportAs: 'layoutDefault',
                template: `
    <div class="alain-default__progress-bar" *ngIf="isFetching"></div>
    <layout-default-header></layout-default-header>
    <div *ngIf="!options.hideAside" class="alain-default__aside">
      <div class="alain-default__aside-inner">
        <ng-container *ngTemplateOutlet="asideUser"></ng-container>
        <ng-container *ngTemplateOutlet="nav"></ng-container>
        <layout-default-nav *ngIf="!nav" class="d-block py-lg"></layout-default-nav>
      </div>
    </div>
    <section class="alain-default__content">
      <ng-container *ngTemplateOutlet="content"></ng-container>
      <ng-content></ng-content>
    </section>
  `
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.NzMessageService }, { type: ɵngcc3.SettingsService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, { options: [{
            type: Input
        }], headerItems: [{
            type: ContentChildren,
            args: [LayoutDefaultHeaderItemComponent, { descendants: false }]
        }], asideUser: [{
            type: Input
        }], nav: [{
            type: Input
        }], content: [{
            type: Input
        }] }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LayoutDefaultHeaderItemTriggerDirective, [{
        type: Directive,
        args: [{
                selector: '[layout-default-header-item-trigger]',
                host: {
                    '[class.alain-default__nav-item]': `true`
                }
            }]
    }], null, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LayoutDefaultHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'layout-default-header',
                template: `
    <ng-template #render let-ls>
      <li *ngFor="let i of ls" [class.hidden-mobile]="i.hidden === 'mobile'" [class.hidden-pc]="i.hidden === 'pc'">
        <ng-container *ngTemplateOutlet="i.host"></ng-container>
      </li>
    </ng-template>
    <div class="alain-default__header-logo" [style.width.px]="options.logoFixWidth">
      <a [routerLink]="options.logoLink" class="alain-default__header-logo-link">
        <img class="alain-default__header-logo-expanded" [attr.src]="options.logoExpanded" [attr.alt]="app.name" />
        <img class="alain-default__header-logo-collapsed" [attr.src]="options.logoCollapsed" [attr.alt]="app.name" />
      </a>
    </div>
    <div class="alain-default__nav-wrap">
      <ul class="alain-default__nav">
        <li *ngIf="!options.hideAside">
          <div class="alain-default__nav-item alain-default__nav-item--collapse" (click)="toggleCollapsed()">
            <i nz-icon [nzType]="collapsedIcon"></i>
          </div>
        </li>
        <ng-template [ngTemplateOutlet]="render" [ngTemplateOutletContext]="{ $implicit: left }"></ng-template>
      </ul>
      <div *ngIf="middle.length > 0" class="alain-default__nav alain-default__nav-middle">
        <ng-container *ngTemplateOutlet="middle[0].host"></ng-container>
      </div>
      <ul class="alain-default__nav">
        <ng-template [ngTemplateOutlet]="render" [ngTemplateOutletContext]="{ $implicit: right }"></ng-template>
      </ul>
    </div>
  `,
                host: {
                    '[class.alain-default__header]': `true`
                },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc3.SettingsService }, { type: LayoutDefaultComponent }, { type: ɵngcc0.ChangeDetectorRef }]; }, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LayoutDefaultNavComponent, [{
        type: Component,
        args: [{
                selector: 'layout-default-nav',
                template: "<ng-template #icon let-i>\n  <ng-container *ngIf=\"i\" [ngSwitch]=\"i.type\">\n    <i\n      *ngSwitchCase=\"'icon'\"\n      class=\"sidebar-nav__item-icon\"\n      nz-icon\n      [nzType]=\"i.value\"\n      [nzTheme]=\"i.theme\"\n      [nzSpin]=\"i.spin\"\n      [nzTwotoneColor]=\"i.twoToneColor\"\n      [nzIconfont]=\"i.iconfont\"\n      [nzRotate]=\"i.rotate\"\n    ></i>\n    <i *ngSwitchCase=\"'iconfont'\" class=\"sidebar-nav__item-icon\" nz-icon [nzIconfont]=\"i.iconfont\"></i>\n    <img *ngSwitchCase=\"'img'\" [src]=\"i.value\" class=\"sidebar-nav__item-icon sidebar-nav__item-img\" />\n    <i *ngSwitchDefault class=\"sidebar-nav__item-icon {{ i.value }}\"></i>\n  </ng-container>\n</ng-template>\n<ng-template #tree let-ls>\n  <ng-container *ngFor=\"let i of ls\">\n    <li\n      *ngIf=\"i._hidden !== true\"\n      class=\"sidebar-nav__item\"\n      [class.sidebar-nav__selected]=\"i._selected\"\n      [class.sidebar-nav__open]=\"i._open\"\n    >\n      <!-- link -->\n      <a\n        *ngIf=\"i.children.length === 0\"\n        (click)=\"to(i)\"\n        [attr.data-id]=\"i._id\"\n        class=\"sidebar-nav__item-link\"\n        [ngClass]=\"{ 'sidebar-nav__item-disabled': i.disabled }\"\n      >\n        <ng-container *ngIf=\"i._needIcon\">\n          <ng-container *ngIf=\"!collapsed\">\n            <ng-template [ngTemplateOutlet]=\"icon\" [ngTemplateOutletContext]=\"{ $implicit: i.icon }\"></ng-template>\n          </ng-container>\n          <span *ngIf=\"collapsed\" nz-tooltip nzTooltipPlacement=\"right\" [nzTooltipTitle]=\"i.text\">\n            <ng-template [ngTemplateOutlet]=\"icon\" [ngTemplateOutletContext]=\"{ $implicit: i.icon }\"></ng-template>\n          </span>\n        </ng-container>\n        <span class=\"sidebar-nav__item-text\" [innerHTML]=\"i._text\" [attr.title]=\"i.text\"></span>\n      </a>\n      <!-- has children link -->\n      <a\n        *ngIf=\"i.children.length > 0\"\n        (click)=\"toggleOpen(i)\"\n        (mouseenter)=\"showSubMenu($event, i)\"\n        class=\"sidebar-nav__item-link\"\n      >\n        <ng-template [ngTemplateOutlet]=\"icon\" [ngTemplateOutletContext]=\"{ $implicit: i.icon }\"></ng-template>\n        <span class=\"sidebar-nav__item-text\" [innerHTML]=\"i._text\" [attr.title]=\"i.text\"></span>\n        <i class=\"sidebar-nav__sub-arrow\"></i>\n      </a>\n      <!-- badge -->\n      <div\n        *ngIf=\"i.badge\"\n        [attr.title]=\"i.badge\"\n        class=\"badge badge-{{ i.badgeStatus }}\"\n        [class.badge-dot]=\"i.badgeDot\"\n      >\n        <em>{{ i.badge }}</em>\n      </div>\n      <ul *ngIf=\"i.children.length > 0\" class=\"sidebar-nav sidebar-nav__sub sidebar-nav__depth{{ i._depth }}\">\n        <ng-template [ngTemplateOutlet]=\"tree\" [ngTemplateOutletContext]=\"{ $implicit: i.children }\"></ng-template>\n      </ul>\n    </li>\n  </ng-container>\n</ng-template>\n<ul class=\"sidebar-nav\">\n  <ng-container *ngFor=\"let group of list\">\n    <li class=\"sidebar-nav__item sidebar-nav__group-title\" *ngIf=\"group.group\">\n      <span [innerHTML]=\"group._text\"></span>\n    </li>\n    <ng-template [ngTemplateOutlet]=\"tree\" [ngTemplateOutletContext]=\"{ $implicit: group.children }\"></ng-template>\n  </ng-container>\n</ul>\n",
                host: {
                    '(click)': '_click()',
                    '(document:click)': '_docClick()'
                },
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc3.MenuService }, { type: ɵngcc3.SettingsService }, { type: ɵngcc1.Router }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.NgZone }, { type: ɵngcc6.DomSanitizer }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [WINDOW]
            }] }, { type: ɵngcc7.Directionality, decorators: [{
                type: Optional
            }] }]; }, { disabledAcl: [{
            type: Input
        }], autoCloseUnderPad: [{
            type: Input
        }], recursivePath: [{
            type: Input
        }], openStrictly: [{
            type: Input
        }], maxLevelIcon: [{
            type: Input
        }], select: [{
            type: Output
        }] }); })();

const COMPONENTS = [
    LayoutDefaultComponent,
    LayoutDefaultNavComponent,
    LayoutDefaultHeaderComponent,
    LayoutDefaultHeaderItemComponent,
    LayoutDefaultHeaderItemTriggerDirective
];
class LayoutDefaultModule {
}
LayoutDefaultModule.ɵfac = function LayoutDefaultModule_Factory(t) { return new (t || LayoutDefaultModule)(); };
LayoutDefaultModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: LayoutDefaultModule });
LayoutDefaultModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            RouterModule,
            NzToolTipModule,
            NzIconModule,
            NzAvatarModule,
            NzDropDownModule,
            NzMessageModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LayoutDefaultModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    NzToolTipModule,
                    NzIconModule,
                    NzAvatarModule,
                    NzDropDownModule,
                    NzMessageModule
                ],
                declarations: COMPONENTS,
                exports: COMPONENTS
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LayoutDefaultModule, { declarations: function () { return [LayoutDefaultComponent, LayoutDefaultNavComponent, LayoutDefaultHeaderComponent, LayoutDefaultHeaderItemComponent, LayoutDefaultHeaderItemTriggerDirective]; }, imports: function () { return [CommonModule,
        RouterModule,
        NzToolTipModule,
        NzIconModule,
        NzAvatarModule,
        NzDropDownModule,
        NzMessageModule]; }, exports: function () { return [LayoutDefaultComponent, LayoutDefaultNavComponent, LayoutDefaultHeaderComponent, LayoutDefaultHeaderItemComponent, LayoutDefaultHeaderItemTriggerDirective]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { LayoutDefaultComponent, LayoutDefaultHeaderComponent, LayoutDefaultHeaderItemComponent, LayoutDefaultHeaderItemTriggerDirective, LayoutDefaultModule, LayoutDefaultNavComponent };
